components/plurals/src/rules/runtime/ast.rs

```diff
@@ -0,0 +1,634 @@
+        // it's hard constructing custom DSTs, we fake a pointer/length construction
+        // eventually we can use the Pointer::Metadata APIs when they stabilize
+        let fake_slice = core::ptr::slice_from_raw_parts(ptr as *const RangeOrValueULE, len_new);
+        let ret = &*(fake_slice as *const Self);
+        debug_assert_eq!(core::mem::size_of_val(ret), core::mem::size_of_val(bytes));
+        ret
+    }
+
+    #[inline]
+    fn as_byte_slice(&self) -> &[u8] {
+        unsafe {
+            core::slice::from_raw_parts(
+                self as *const Self as *const u8,
+                core::mem::size_of_val(self),
+            )
+        }
+    }
+
+    //     what it should do is attempt to parse the first 4 bytes as a u32::ULE (POU<4>), and the remaining bytes as a ZV<RangeOrValueULE>
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let bits = bytes[0]; // XXX: Validate those bits
+        <PlainOldULE<4> as ULE>::validate_byte_slice(&bytes[1..5]).map_err(|_| "foo")?;
+        let remaining = &bytes[5..];
+        RangeOrValueULE::validate_byte_slice(remaining).map_err(|_| "foo")?;
+        Ok(())
+    }
+}
+
+unsafe impl EncodeAsVarULE<RelationULE> for Relation<'_> {
+    fn encode_var_ule_as_slices<R>(&self, cb: impl FnOnce(&[&[u8]]) -> R) -> R {
+        let encoded =
+            RelationULE::encode_andor_polarity_operand(self.and_or, self.polarity, self.operand);
+        cb(&[
+            &[encoded],
+            PlainOldULE::<4>::as_byte_slice(&[self.modulo.as_unaligned()]),
+            self.range_list.as_bytes(),
+        ])
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct RangeOrValueULE(PlainOldULE<8>);
+
+unsafe impl ULE for RangeOrValueULE {
+    type Error = zerovec::ule::ULEError<core::convert::Infallible>;
+
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        PlainOldULE::<8>::validate_byte_slice(bytes)
+    }
+}
+
+impl AsULE for RangeOrValue {
+    type ULE = RangeOrValueULE;
+
+    #[inline]
+    fn as_unaligned(self) -> Self::ULE {
+        let mut result = [0; 8];
+        match self {
+            Self::Range(start, end) => {
+                let start_bytes = start.to_be_bytes();
+                let end_bytes = end.to_be_bytes();
+                result[..4].copy_from_slice(&start_bytes);
+                result[4..].copy_from_slice(&end_bytes);
+                RangeOrValueULE(result.into())
+            }
+            Self::Value(idx) => {
+                let bytes = idx.to_be_bytes();
+                result[..4].copy_from_slice(&bytes);
+                result[4..].copy_from_slice(&bytes);
+                RangeOrValueULE(result.into())
+            }
+        }
+    }
+
+    #[inline]
+    fn from_unaligned(unaligned: Self::ULE) -> Self {
+        let b = unaligned.0 .0;
+        let start = u32::from_be_bytes([b[0], b[1], b[2], b[3]]);
+        let end = u32::from_be_bytes([b[4], b[5], b[6], b[7]]);
```

Comment:
Thought: This isn't as nice as it could be.  If you take my suggestion to use `PairULE`, it should clean this up.